atom/browser/api/event_subscriber.cc

```diff
@@ -1,101 +1,100 @@
+  auto weak_self = weak_factory_.GetWeakPtr();
+  On("data", base::Bind(&EventSubscriber::OnData, weak_self));
+  On("end", base::Bind(&EventSubscriber::OnEnd, weak_self));
+  On("error", base::Bind(&EventSubscriber::OnError, weak_self));
 }
 
-EventSubscriberBase::~EventSubscriberBase() {
-  if (!isolate_) {
-    return;
-  }
+EventSubscriber::~EventSubscriber() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
   RemoveAllListeners();
-  emitter_.Reset();
-  DCHECK_EQ(js_handlers_.size(), 0u);
 }
 
-void EventSubscriberBase::On(const std::string& event_name) {
-  DCHECK(js_handlers_.find(event_name) == js_handlers_.end());
+void EventSubscriber::On(const std::string& event, EventCallback&& callback) {  // NOLINT
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  DCHECK(js_handlers_.find(event) == js_handlers_.end());
+
   v8::Locker locker(isolate_);
   v8::Isolate::Scope isolate_scope(isolate_);
   v8::HandleScope handle_scope(isolate_);
-  auto fn_template = g_cached_template.Get(isolate_);
-  auto event = mate::StringToV8(isolate_, event_name);
-  auto js_handler_data = new JSHandlerData(isolate_, this);
-  v8::Local<v8::Value> fn = internal::BindFunctionWith(
-      isolate_, isolate_->GetCurrentContext(), fn_template->GetFunction(),
-      js_handler_data->handle_.Get(isolate_), event);
-  js_handlers_.insert(
-      std::make_pair(event_name, v8::Global<v8::Value>(isolate_, fn)));
-  internal::ValueVector converted_args = {event, fn};
-  internal::CallMethodWithArgs(isolate_, emitter_.Get(isolate_), "on",
-                               &converted_args);
+
+  // emitter.on(event, EventEmitted)
+  auto fn = CallbackToV8(isolate_, callback);
+  js_handlers_[event] = v8::Global<v8::Value>(isolate_, fn);
+  internal::ValueVector args = { StringToV8(isolate_, event), fn };
+  internal::CallMethodWithArgs(isolate_, emitter_.Get(isolate_), "on", &args);
 }
 
-void EventSubscriberBase::Off(const std::string& event_name) {
+void EventSubscriber::Off(const std::string& event) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  DCHECK(js_handlers_.find(event) != js_handlers_.end());
+
   v8::Locker locker(isolate_);
   v8::Isolate::Scope isolate_scope(isolate_);
   v8::HandleScope handle_scope(isolate_);
-  auto js_handler = js_handlers_.find(event_name);
+  auto js_handler = js_handlers_.find(event);
   DCHECK(js_handler != js_handlers_.end());
   RemoveListener(js_handler);
 }
 
-void EventSubscriberBase::RemoveAllListeners() {
+void EventSubscriber::OnData(mate::Arguments* args) {
+  v8::Local<v8::Value> buf;
+  args->GetNext(&buf);
+  if (!node::Buffer::HasInstance(buf)) {
+    args->ThrowError("data must be Buffer");
+    return;
+  }
+
+  // Copy the data from Buffer.
+  // TODO(zcbenz): Is there a way to move the data?
+  const char* data = node::Buffer::Data(buf);
+  size_t length = node::Buffer::Length(buf);
+  std::vector<char> buffer(length);
+  buffer.assign(data, data + length);
```

Comment:
Tip: Initialize the vector with the source buffer to avoid zeroing the buffer first only to replace the content with actual data.
```c++
std::vector<char> buffer(data, length);
```